nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,354 @@
+
+class DartsLayerChoice(nn.Module):
+    def __init__(self, layer_choice):
+        super(DartsLayerChoice, self).__init__()
+        self.name = layer_choice.label
+        self.op_choices = nn.ModuleDict(OrderedDict([(name, layer_choice[name]) for name in layer_choice.names]))
+        self.alpha = nn.Parameter(torch.randn(len(self.op_choices)) * 1e-3)
+
+    def forward(self, *args, **kwargs):
+        op_results = torch.stack([op(*args, **kwargs) for op in self.op_choices.values()])
+        alpha_shape = [-1] + [1] * (len(op_results.size()) - 1)
+        return torch.sum(op_results * F.softmax(self.alpha, -1).view(*alpha_shape), 0)
+
+    def parameters(self):
+        for _, p in self.named_parameters():
+            yield p
+
+    def named_parameters(self, recurse):
+        for name, p in super(DartsLayerChoice, self).named_parameters():
+            if name == 'alpha':
+                continue
+            yield name, p
+
+    def export(self):
+        return list(self.op_choices.keys())[torch.argmax(self.alpha).item()]
+    
+    def r_forward(self, x):
+        return self.op_choices.values()[0](x)
+
+class DartsInputChoice(nn.Module):
+    def __init__(self, input_choice):
+        super(DartsInputChoice, self).__init__()
+        self.name = input_choice.label
+        self.alpha = nn.Parameter(torch.randn(input_choice.n_candidates) * 1e-3)
+        self.n_chosen = input_choice.n_chosen or 1
+
+    def forward(self, inputs):
+        inputs = torch.stack(inputs)
+        alpha_shape = [-1] + [1] * (len(inputs.size()) - 1)
+        return torch.sum(inputs * F.softmax(self.alpha, -1).view(*alpha_shape), 0)
+
+    def parameters(self):
+        for _, p in self.named_parameters():
+            yield p
+
+    def named_parameters(self, recurse):
+        for name, p in super(DartsInputChoice, self).named_parameters():
+            if name == 'alpha':
+                continue
+            yield name, p
+
+    def export(self):
+        return torch.argsort(-self.alpha).cpu().numpy().tolist()[:self.n_chosen]
+    
+    def r_forward(self, x):
+        return super().r_forward(x)
+
+DARTS_REPLACE_DICT = {
+    LayerChoice : DartsLayerChoice,
+    InputChoice : DartsInputChoice
+}
+
+class DartsModel(BaseOneShotLightningModule):
+    '''
+    The DARTS model. In each iteration, there are 2 training phases. The phase 1 is architecture
+    step, in which the model parameters are frozen and the alphas are optimized. The phase 2 is model
+    step, in wchich the alphas are frozen and the model parameters are optimized. In this step, the
+    out put of each choice are sumed up with respect to their alpha value. The result of DARTS
+    is argmax in alpha.
+    The DARTS Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
+
+    Reference
+    ----------
+    [1]H. Liu, K. Simonyan, and Y. Yang, “DARTS: Differentiable Architecture Search,” presented at the
```

Comment:
Missing space after `[1]`.
Missing `..` before `[1]`.
You can use `[darts]` instead of `[1]`.
Add indent in the second line.